File: 05_expense_tracker/python/solution.py
expenses = []

def add_expense(amount, category):
    expenses.append((amount, category))
    print("Expense added.")

def show_total():
    total = sum(amount for amount, _ in expenses)
    print(f"Total expenses: {total:.2f}")

def show_by_category():
    category_totals = {}
    for amount, category in expenses:
        category_totals[category] = category_totals.get(category, 0) + amount

    for category, total in category_totals.items():
        print(f"{category}: {total:.2f}")

def menu():
    while True:
        print("\n1. Add Expense")
        print("2. View Total")
        print("3. View by Category")
        print("4. Exit")
        choice = input("Choose option: ")

        if choice == "1":
            try:
                amount = float(input("Amount: "))
                category = input("Category: ")
                add_expense(amount, category)
            except ValueError:
                print("Invalid amount.")
        elif choice == "2":
            show_total()
        elif choice == "3":
            show_by_category()
        elif choice == "4":
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    menu()

File: 05_expense_tracker/java/Solution.java
import java.util.*;

public class Solution {

    private static final List<Expense> expenses = new ArrayList<>();

    private static class Expense {
        double amount;
        String category;

        Expense(double amount, String category) {
            this.amount = amount;
            this.category = category;
        }
    }

    private static void addExpense(double amount, String category) {
        expenses.add(new Expense(amount, category));
        System.out.println("Expense added.");
    }

    private static void showTotal() {
        double total = expenses.stream().mapToDouble(e -> e.amount).sum();
        System.out.printf("Total expenses: %.2f%n", total);
    }

    private static void showByCategory() {
        Map<String, Double> map = new HashMap<>();
        for (Expense e : expenses)
            map.put(e.category, map.getOrDefault(e.category, 0.0) + e.amount);

        map.forEach((k, v) ->
                System.out.printf("%s: %.2f%n", k, v));
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n1. Add Expense");
            System.out.println("2. View Total");
            System.out.println("3. View by Category");
            System.out.println("4. Exit");
            System.out.print("Choice: ");
            String choice = sc.nextLine();

            switch (choice) {
                case "1":
                    try {
                        System.out.print("Amount: ");
                        double amount = Double.parseDouble(sc.nextLine());
                        System.out.print("Category: ");
                        String category = sc.nextLine();
                        addExpense(amount, category);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid amount.");
                    }
                    break;
                case "2":
                    showTotal();
                    break;
                case "3":
                    showByCategory();
                    break;
                case "4":
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

